### PR TITLE
Date time from immutable

### PR DESCRIPTION
**Pull request by tpetracca**
Wanted a way for people to have an idea for what point in time their backup is for.
@clockfort

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/InMemoryPuncherStore.java [hunk 1]
```diff
@@ -37,7 +37,7 @@ public static InMemoryPuncherStore create() {
     private InMemoryPuncherStore() {//
     }
 
-    private final SortedMap<Long, Long> map = makeMap();
+    private final TreeMap<Long, Long> map = makeMap();
```
Comment (schlosna) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/InMemoryPuncherStore.java]: `NavigableMap`?

File: atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/TimestampCommand.java [hunk 2]
```diff
@@ -31,17 +35,35 @@
     		description = "Get the current immutable timestamp")
     private boolean immutable;
     
+    @Option(name = {"-d", "--date-time"},
+            description = "Return the earliest real point in time at which the immutable timestamp" +
```
Comment (clockfort) [inline atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/TimestampCommand.java]: I was confused by 'real point in time' and would prefer 'wallclock time' but idk if that's just me

Comment (clockfort) [inline atlasdb-cli/src/main/java/com/palantir/atlasdb/cli/command/TimestampCommand.java]: And with async puncher I think what we're retrieving with your logic here is "approximate wallclock time"

**Comment by clockfort** [comment]
generally lgtm and seems like a feature someone could want to use to sanity check a backup or something, as you said.

**Comment by tpetracca** [comment]
testing is hard.

**Comment by tpetracca** [comment]
I'm a noob, and don't really know how Dagger or DI generally works.  If there's a way to make the whole providing a cleaner thing optional (and have it use a NoOpCleaner if you don't provide) then let me know.

**Comment by schlosna** [comment]
:+1:

**Comment by tpetracca** [comment]
Also I can't actually find the error in that verify log?

**Comment by tpetracca** [comment]
I think https://github.com/palantir/atlasdb/pull/271 may have caused this.  The last test it gets to is :lock-impl:test before dying on a timeout: Still running (20 of 20): ./gradlew --max-workers 2 check --stacktrace
Timeout (20 minutes) reached. Terminating "./gradlew --max-workers 2 check --stacktrace".

Could also just be that the test is still flaky.  Someone want to kick this verify?

**Comment by clockfort** [comment]
Verify kicked and commit merged.

**Pull request merged**